### ISSUE TITLE
Prevent creating page that matches allowed UrlSegments

### ISSUE THREAD

**Issue by adrianbj** [processwire/processwire-issues#1333]
### Short description of the issue

Not sure if this should be a feature request, but it does result in the possibility of breaking the site, so maybe it's a bug?

The issue is if you define allowed segments, eg:

```
regex:^people/[A-Za-z0-9-]+$
```

and then create a child page with name = `people` then the page will 404.

I know I could prevent this with a hook on Page Add, but I feel like the core page creation should try to prevent this, because you never know what clients will do :)

### Expected behavior

Prevent creating a page with a name that matches a defined url segment.

**Comment by teppokoivula** [comment]
Shouldn't real pages always take precedence over URL segments? That's literally what the docs say:

> Should there be a child page that has the same name as a URL segment that the parent page's template is looking for, the pages in the tree always have precedence. Meaning, URL segments only apply if the requested URL did not resolve to an actual page.

I use allowed segment settings very rarely, preferring code validation instead, so this behaviour is not particularly familiar to me. If it works as you've described, i.e. allowed URL segment setting makes real pages unreachable, that doesn't seem to be in line with the docs, and thus sounds like a bug to me.

TL;DR: instead of preventing creation of a page, I think it would be better to give the page precedence here. If Page creation should be disabled instead, that could indeed be done with a hook :)

**Comment by adrianbj** [comment]
~~I am not sure that real pages should always take precedence~~ - when you have a multi-path urlsegment, the first segment's "name" is key to an entirely different piece of logic being implemented and a real page shouldn't be able to break that, so I still think that it would be a useful addition to prevent breaking of an entire key part of the site simply by creating a page - maybe this should be case-by-case with hooks, but if the core could handle it, I think that would be nice.

I do often use the allow urlesegments definition, but what is forcing me to use it on this project is because I am using this module: https://processwire.com/modules/page-hit-counter/ which requires it. The last page of its forum support thread discusses why it needs it.

Note that that module wouldn't need this if PW had a dedicated endpoint that modules could make use of - see the discussion here: https://processwire.com/talk/topic/25145-routing-feature-custom-endpoints-without-hooking-pagenotfound404/

EDIT: Obviously real pages do need to take precedence, but I think blocking creation of a real page that would break a urlsegment should be prevented. Maybe it's too hard to cover all scenarios so maybe it does need to be handled case-by-case via a Page Add hook, but if it could be handled via the core, that would be awesome.

**Comment by ryancramerdesign** [comment]
@adrianbj It sounds interesting, and while I don't think I can support that feature at present, I will keep it in mind for the future. But wanted to mention that in your regex, you've got a hyphen as the last character in your group. Since hyphens have a special meaning (range) in character groups, I think that PCRE requires a hyphen to be the first character in the group in order for it to work, so I think this is what you want: `regex:^people/[-A-Za-z0-9]+$`

**Comment by adrianbj** [comment]
@ryancramerdesign - thanks for thinking about it @ryancramerdesign - while it might be an unlikely scenario, it's definitely a big problem if it does happen.

Regarding the regex, I don't believe it makes a difference. You can see it works as is here: https://www.phpliveregex.com/p/Agb

and to quote an SO comment: "the final dash will not be interpreted as a range operator (since the range has no end)."

**Comment by ryancramerdesign** [comment]
@adrianbj Okay good to know. When I learned regex's it was like 20 years ago and back then at least I think you could only have a hyphen/dash at the beginning. But it makes sense that it would support it at either beginning or end.

**Comment by adrianbj** [comment]
The other option is to escape it. Maybe some regex engines still need that, but it seems PCRE is flexible on this front.

**Comment by adrianbj** [comment]
Hey @ryancramerdesign - when I posed to Claude I got this response. Apologies if you already tried and got a similar response. I can certainly use the suggested hook, but the other options here should also work. I haven't tested the code below for Option 2 or  Option 3.

But I have tested its attempt at Option 1 which seems to be working as expected with the added advantage of allow the real page to still work. You can see it here: https://github.com/processwire/processwire/compare/master...adrianbj:processwire:issue-1333-fix-option-1#diff-c9829ce50a1a092b2795d0a3f34d18a87ea25197af6988e598659f952e1c56a7

## Claude analysis

 The Problem                                                                                                                                                                                                      
                                                                                                                                                                                                                   
  When a parent template defines multi-path URL segments like regex:^people/[A-Za-z0-9-]+$, and someone creates a child page named people, the path resolution in PagesPathFinder::getPagesRow() uses LEFT JOINs   
  that match the real child page first. This means:                                                                                                                                                                
                                                                                                                                                                                                                   
  - /parent/people/john-doe/ resolves people as a real child page (not a URL segment)                                                                                                                              
  - john-doe becomes a URL segment on the people page's template
  - If that template doesn't support URL segments → 404                                                                                                                                                            
                                                                                                                                                                                                                 
  The parent's URL segment pattern people/john-doe never gets evaluated because the first part matched a real page.                                                                                                
                                                                                                                                                                                                                 
  Fix Options                                                                                                                                                                                                      
                                                                                                                                                                                                                 
  Option 1: Fallback in PagesPathFinder (best fix)                                                                                                                                                                 
   
  In PagesPathFinder::applyResultTemplate() (~line 688), when URL segments fail validation on the resolved page, fall back to treating more of the path as URL segments on a parent page. Specifically, in         
  applyPagesRow() at line 402-428 of wire/core/PagesPathFinder.php:                                                                                                                                              
                                                                                                                                                                                                                   
  After the initial resolution, if URL segments are present and fail template validation, retry by "rewinding" — treat the last matched page name as a URL segment on its parent instead. This would check if      
  people/john-doe validates as URL segments on the parent template.
                                                                                                                                                                                                                   
  Option 2: Prevent page creation (what the issue asks for)                                                                                                                                                      

  Add validation in Pages::___save() (wire/core/Pages.php:856) or via a Pages::saveReady hook. Before saving, check if the parent page's template has URL segment rules, and if the new page's name would conflict 
  with the first segment of any rule:
 ```                                                                                                                                                                                                                  
  // In a Pages::saveReady hook or in ___save()                                                                                                                                                                  
  $parent = $page->parent;                                                                                                                                                                                         
  $parentTemplate = $parent->template;                                                                                                                                                                             
  $rules = $parentTemplate->urlSegments();                                                                                                                                                                         
                                                                                                                                                                                                                   
  if(is_array($rules)) {                                                                                                                                                                                         
      $pageName = $page->name;
      foreach($rules as $rule) {                                                                                                                                                                                   
          if(stripos($rule, 'regex:') === 0) {
              $regex = '{' . trim(substr($rule, 6)) . '}';                                                                                                                                                         
              // Check if any URL segment pattern starts with this page name                                                                                                                                     
              if(preg_match($regex, "$pageName/test") || preg_match($regex, $pageName)) {                                                                                                                          
                  // Warn or prevent save                                                                                                                                                                          
                  throw new WireException("Page name '$pageName' conflicts with URL segment rule on parent template");                                                                                             
              }                                                                                                                                                                                                    
          } else if(strpos($rule, '/') !== false) {                                                                                                                                                              
              $firstSegment = explode('/', $rule)[0];                                                                                                                                                              
              if($pageName === $firstSegment) {                                                                                                                                                                  
                  throw new WireException("Page name '$pageName' conflicts with URL segment '$rule' on parent template");                                                                                          
              }                                                                                                                                                                                                  
          }                                                                                                                                                                                                        
      }                                                                                                                                                                                                          
  }
```                                                                                                                                                                                                                   
  Option 3: Hook-based solution (simplest, user-land)                                                                                                                                                              
                                                                                                                                                                                                                   
  The easiest approach without modifying the core — add a hook in site/ready.php:                                                                                                                                  
```                                                                                                                                                                                                                 
  $pages->addHookBefore('Pages::saveReady', function(HookEvent $event) {                                                                                                                                           
      $page = $event->arguments(0);                                                                                                                                                                                
      if(!$page->parent || !$page->parent->id) return;
                                                                                                                                                                                                                   
      $parentTemplate = $page->parent->template;                                                                                                                                                                 
      $rules = $parentTemplate->urlSegments();                                                                                                                                                                     
      if(!is_array($rules)) return;                                                                                                                                                                              

      foreach($rules as $rule) {                                                                                                                                                                                   
          if(stripos($rule, 'regex:') === 0) {
              $regex = '{' . trim(substr($rule, 6)) . '}';                                                                                                                                                         
              if(preg_match($regex, $page->name) || preg_match($regex, "{$page->name}/test")) {                                                                                                                  
                  throw new WireException("Cannot create page '{$page->name}' — conflicts with URL segment pattern on parent template '{$parentTemplate->name}'");                                                 
              }                                                                                                                                                                                                    
          } elseif($page->name === explode('/', $rule)[0]) {                                                                                                                                                       
              throw new WireException("Cannot create page '{$page->name}' — conflicts with URL segment on parent template '{$parentTemplate->name}'");                                                             
          }                                                                                                                                                                                                        
      }                                                                                                                                                                                                            
  });                                                                                                                                                                                                              
 ```                                                                                                                                                                                                        
  Recommendation

  Option 1 is the most robust core fix — it respects the documented behavior ("real pages take precedence") while gracefully falling back when the resolved page can't handle the remaining URL segments. The key  
  change would be in PagesPathFinder.php to retry resolution when applyResultTemplate() fails due to urlSegmentsOFF or urlSegmentsBAD.
